### DISCUSSION TITLE
axum-login with 'Sign in with Google for Web'

### DISCUSSION THREAD

**Discussion by danburkert** [maxcountryman/axum-login#244] [Q&A]
Hello,

I'm trying to integrate `axum-login` with an application which uses [Sign in with Google for Web](https://developers.google.com/identity/gsi/web/) (GSI) to handle authentication.  GSI is similar to OIDC, in that an ID token is provided to the application while authenticating a user.  I'm struggling to understand how to implement the `AuthUser::session_auth_hash` functionality.

The oauth2 example in the repo appears to extract an access token secret from the JWT during authentication, which is then stored into the database.  GSI doesn't have an access token as such, so this isn't directly translatable. Additionally, I suspect that this implementation doesn't allow a user to be actively signed in to the application on two different devices (say a laptop and mobile), since the JWTs on each device would be different, and the second to authenticate would effectively invalidate the `session_auth_hash` of the first and thus log it out? I'd like to support this usecase if possible.  Furthermore, this example makes me a bit uneasy because it's storing an un-hashed credential directly in the database, which is dangerous in the event of a database leak. I don't see an obvious reason for storing the access token, since it would seem to be functionally equivalent to randomly generating a 'hash'  directly following authentication, and storing that in the database?

Thanks,

- Dan

**Comment by maxcountryman** [comment]
I'm not really sure what your question is.

**Reply by danburkert** [reply]
When using GSI or OIDC for authentication, how should `AuthUser::session_auth_hash` be implemented?

**Reply by maxcountryman** [reply]
That's up to you. It's a semantic you define for your use case.

**Reply by danburkert** [reply]
OK, I suppose I'm asking about a typical web application usecase, if such a thing exists.

In my case, I'm coalescing a design around a relational database with separate `users` and `sessions` tables which would allow a user to have multiple active sessions (for different devices or browsers).  In such a scenario, I believe it's necessary to treat the `AuthUser` implementation as not (only) a user, but the combination of a user + session.

The only hangup I'm having about such a design, which I hinted at in the original question, is that a randomly generated 16-byte session-ID used as the `session_auth_hash` value must be stored un-hashed in the database. I've found numerous recommendations that such session IDs not be stored un-hashed in a database, e.g: https://security.stackexchange.com/questions/63435/why-use-an-authentication-token-instead-of-the-username-password-per-request/63438#63438.  To allow the session-ID to be hashed, `axum-login` would need to be extended to allow a transformation function over the bytes stored in the cookie before comparing to the value in the database.  Is that a usecase you want to support with axum-login?  I'm unsure if this is just missing functionality, or perhaps I'm entirely misunderstanding the intention behind `session_auth_hash`.

**Reply by maxcountryman** [reply]
Again, it's up to your implementation to decide such things. The name of the method implies you should be producing a secure hash, of course.

Such a hash should be safe to store (just like you might store the Argon2 hash of a password), but again you'll need to ensure that's true.

Nothing stops you from using a secure hashing method in the implementation itself.

> I believe it's necessary to treat the AuthUser implementation as not (only) a user, but the combination of a user + session.

Personally I wouldn't do things this way: instead I would have a notion of `User` (implementing `AuthUser` for this) and `Account`, where users are one-to-many with accounts.

**Reply by danburkert** [reply]
> The name of the method implies you should be producing a secure hash, of course.

A hash of what, though?  In the GSI/OIDC case there is no equivalent of a password.  That's why the design I outlined includes a randomly generated session-ID, aka an access token.

> Such a hash should be safe to store (just like you might store the Argon2 hash of a password), but again you'll need to ensure that's true.

There is a difference between storing the Argon2 hash of a password, and a randomly generated access token - an attacker with access to the database can not use the Argon2 hash of a password to hijack a session or otherwise authenticate.  However if an access token leaks, it's trivial to use that token to authenticate.

> Personally I wouldn't do things this way: instead I would have a notion of User (implementing AuthUser for this) and Account, where users are one-to-many with accounts.

Can you expand on what an account does in this scenario?  I'm not familiar with a distinction between a user and an account in the context of a web application.

**Reply by maxcountryman** [reply]
I'm available to help answer more involved design considerations as a consultant. Feel free to shoot me an email and we can discuss details of what that might look like.

If that's not an option for you, then I'd invite you to explore some of the prior discussion around ways to approach this (I've previously mentioned designs for e.g. multiple clients per OIDC session). You may also want to take a look at the Django middleware this is based on.

**Reply by danburkert** [reply]
OK, thanks for your help!
